### PR TITLE
mc-9534 provide methods so that subclasses can provide their own meta…

### PR DESCRIPTION
**Pull request by aaronforshaw**
…data namespaces

Necessary for mc-9533 and mc-9534 (add profiles to SqlServer and PostgreSQL plugins).

Because profiles require distinct metadata namespaces, and because we want to apply profiles at the database, schema, table and column levels, the plugins which provide profiles need to be able to set their own separate metadata namespaces for database, schema, table and column.

This PR adds methods which keep the default behaviour unchanged, but allow relevant plugins to override metadata namespaces for profiles.

### CODE REVIEW

**Comment by olliefreeman** [comment]
why do we need differect namespaces for each of the levels, we currently store the metadata which would end up in the profile all using the same namespace, and as profiles collect metadata based on the object being assessed what need is there for a DC to use a different NS to that of a DE etc??

also the intent here really is that if you want to export a DM to a database of sql server origin then you apply the sqlserver profile, if you want to export the same DM to a postgres DB then you also apply the postgres profile. Both of these importers/exporters already have very clearly different MD NSs.

I don't believe that we should allow 2 profiles in one DM to export to 2 sqlserver instances with different configs.

Basically i dont understand why we need these changes.

**Comment by aaronforshaw** [comment]
> why do we need differect namespaces for each of the levels, we currently store the metadata which would end up in the profile all using the same namespace, and as profiles collect metadata based on the object being assessed what need is there for a DC to use a different NS to that of a DE etc??
> 
> also the intent here really is that if you want to export a DM to a database of sql server origin then you apply the sqlserver profile, if you want to export the same DM to a postgres DB then you also apply the postgres profile. Both of these importers/exporters already have very clearly different MD NSs.
> 
> I don't believe that we should allow 2 profiles in one DM to export to 2 sqlserver instances with different configs.
> 
> Basically i dont understand why we need these changes.

@olliefreeman I started off by using the same metadata namespace at the DM, DC and DE levels. But this didn't work because the ProfileController returned at random a wrong profile, for example it would return the DC profile for the DM, DC and DE. I assumed this was because in EmptyJsonProfileFactory the profile structure is keyed by metadataNamespace:

```
    private String getOrLoadProfileStructure(JsonProfileProviderService jsonProfileProviderService) {
        if (!loadedProfileJsonStructures[jsonProfileProviderService.metadataNamespace]) {
            loadedProfileJsonStructures[jsonProfileProviderService.metadataNamespace] = loadNewProfileStructure(jsonProfileProviderService)
        }
        loadedProfileJsonStructures[jsonProfileProviderService.metadataNamespace]
    }
```

and found that by using separate metadata namespaces, it works OK.

**Comment by olliefreeman** [comment]
Ah right, I see, its more about what Profile Json file is loaded for the MD NS rather than how the data is stored in the database. Fair enough

**Review by olliefreeman — APPROVED** [review]
